Enhancement: Expose Response Headers and Rate Limit Metrics, Efficiency Logging

mecfletcher opened the pull request:
This pull request introduces an array of updates aimed at providing greater visibility for response-level information and optimizing processing efficiency. Key enhancement details include the following:

Additions:
setResponseHeaders and getResponseHeaders methods have been implemented to allow the storage of response headers.
setRateLimitInfo, getRateLimitInfo, and hydrateRateLimitInfo methods have been implemented to provide visibility of rate-limiting information.
setProcessingMs, getProcessingMs, and hydrateProcessingMs methods have been implemented to monitor processing time in milliseconds.
setHeaderFunction method has been implemented to manage header formatting and assignment.
Refactoring:
Existing methods have been refactored to accommodate the setHeaderFunction method and to hydrate the rate limiting and processing info after every request.
These changes will assist in monitoring usage, complying with rate-limits, and optimizing processing speed. Further testing and feedback are appreciated.

Commit log:
Provide access to responseHeaders
Add processingMs and cleanup
Cleanup sub-indexing
Implement isset
Fix indexing
Refactor to setHeaderMethod and error_log for test
Fix key values for rate limit headers
Cleanup
Fix default values to avoid confusion

From the OpenAi Docs:
<img width="774" alt="Screenshot 2023-11-14 at 2 25 17 PM" src="https://github.com/orhanerday/open-ai/assets/45575335/66b8ad22-7c15-42a4-8408-e43ecd232835">

maximehinnekens commented:
This is highly needed!

orhanerday commented:
Sorry for the delay, @mecfletcher. 

Could you please explain to me how I can test this? Do you have any usage examples?

mecfletcher commented:
@orhanerday I apologize for the delay on my end, I've added a usage section to the PR description and added some better formatting! Let me know if you have any other questions.
